hadoop-hdds/docs/content/design/leader-execution/leader-execution.md

```diff
@@ -0,0 +1,477 @@
+
+
+### Gatekeeper
+Gatekeeper act as entry point for request execution. Its function is:
+1. orchestrate the execution flow
+2. granular locking
+3. execution of request
+4. validate om state like upgrade
+5. update metrics and return response
+6. handle client retry / replay of request
+7. managed index generation (remove dependency with ratis index for objectId)
+
+### Executor
+This prepares context for execution, process the request, communicate to all nodes for db changes via ratis and clearing up any cache.
+
+### Batching (Ratis request)
+All requests executed in parallel are batched and send as single request to other nodes. This helps improve performance over network with batching.
+
+Batching of Request:
+- Request 1..n are executed and db changes are identified and added to queue (and request will be waiting for update via ratis over Future waiting)
+- Batcher will retrieve Request 1..n and db changes, merge those request to single Ratis Request message
+- Send Merged Request message to all nodes via ratis and receive reply
+- Batcher will reply to each request 1..n with db update success notifying future object of each request.
+
+There are multiple batchers waiting over queue,
+- As soon as queue have entry, and the batcher is available, it will pick all request from queue for processing
+- batcher will be un-available when its processing the batch, i.e. merge request and send to ratis and then waiting for reply
+
+As performance Test result, Number of batcher with "5->8" performed the best. 
+- Higher number of batcher reduces the effective batching, and performance reduces
+- Lower number of batcher reduces throughput as more request will be waiting for ratis response
+
+### Apply Transaction (via ratis at all nodes)
+With new flow as change,
+- all nodes during ratis apply transaction will just only update the DB for changes.
+- there will not be any double buffer and all changes will be flushed to db immediately.
+- there will be few specific action like snapshot creation of db, upgrade handling which will be done at node.
+- And response to client will be returned after the apply transaction is success to the nodes in quorum.
+
+## Description
+
+### Index generation
+
+In old flow, ratis index is used for `object Id` of key and `update Id` for key update.
+For new flow, it will not depend on ratis index, but will have its own **`managed index`**.
+
+Index initialization / update:
+- First time startup: 0
+- On restart (leader): last preserved index + 1
+- On Switch over: last index + 1
+- Request execution: index + 1
+- Upgrade: Last Ratis index + 1  (This is only for existing cluster during upgrade)
+
+Om is going to maintain `IndexGenerator` which will maintain,
+- index: Atomic long, which will generate new index using incrementAndGet()
+- commitIndex: This is saved index in DB at follower and leader. This will be used to update `index` when a follower becomes Leader on ratis's notifyLeaderChanged.
```

Comment:
If a follower's commitIndex is behind two other nodes, does the follower have a chance to be selected as the leader?

Reply:
Here if follower's commit index is behind than other two nodes, than implicitly, ratis index will be also behind as those transactions are not executed. So ratis will try to select leader with most current index as I know.

From behavior perspective, there is no change as commit index change is linked to ratis index changes implicitly. Leader election and data which is given preference will be dependent on ratis logic.

No change will be observed with the above change, just object generation index is separated out.